ironfish/src/rpc/clients/socketClient.ts

```diff
@@ -349,6 +350,9 @@ export abstract class RpcSocketClient extends RpcClient {
 
   protected onError(error: unknown): void {
     this.logger.error(ErrorUtils.renderError(error))
+    for (const request of this.pending.values()) {
```

Comment:
Can you explain what this is meant to try to fix?

This is already handled in `onClientClose` in the case that the connection is closed unexpectedly. Otherwise, onError can be called if one request fails, but the other requests shouldn't be rejected in that case of one error. I think the way this is intended, an error doesn't necessarily mean the end of valid responses for all pending requests.

Reply:
In the Unix world, there are only two errors that I know that may be retried with recv()/read()/send()/write() and friends, when working with sockets: 
- EAGAIN, when using nonblocking sockets (which is not the case here, I think),
- EINTR, when not using SA_RESTART when setting up signal handlers (which again I don't think can occur here).

All other errors, to my knowledge, are errors that leave the socket or the HTTP request in an invalid state that cannot be recovered. Some common scenarios of when this may occur are: the connection is closed abruptly (without receiving an RST packet from the source), routing tables change somewhere in the path and the peer becomes unreachable, the maximum number of TCP retries is hit, and so on. In all those cases, a "connection closed" event may not be fired (because technically the connection did not get closed on the TCP/IP layer), and in all those cases it won't be possible to reliably know what should be done next with the connection.

For example: suppose that the client sends the HTTP headers for a request on a TCP packet. Routing changes, the destination becomes unreachable, and an error is thrown. At that point, even if the routing is corrected, it won't be possible to recover the situation: it should in theory be possible to resend the HTTP headers in a new TCP packet, but from the Node.js exception we don't have enough information to know what TCP packet needs to be re-transmitted, nor we have the functionality from the HTTP library to do the re-transmission. Moreover, it is actually quite unlikely that such errors will recover themselves in a short period of time, and so any re-transmission logic may actually end up hanging there forever.

In case of error, it is way more reliable to just discard the connection, establish a new one, and retry all in-flight HTTP requests. This PR doesn't really do any of that: it just propagates the error to the upper layer in the hope that those layers have the correct logic to handle this situation.

So, in short, errors are (generally speaking) not retry-able, and errors may not result in a "connection closed" event.

Reply:
Note that, in the scenario that motivated this PR, we have an ECONNRESET ("connection reset by peer") returned during a send() call, so here technically a connection was closed, but for some reason Node.js does not want treat this situation as a "connection closed", but rather wants to treat is as an error condition.

Reply:
Just to add another example of why continuing when an error is present (if possible) may result in unintended consequences:

1. http headers for request A are sent
2. http body is sent, but an error occurs so it's never received
3. http headers for request B are sent

In this situation, the server will think that the headers for request B are the body of request A